# Lecture 2 Data Types in Python

The lecture contains the following subsections:
- [2.1 Introduction](#section1)
- [2.2 Numbers](#section2)
- [2.3 Strings](#section3)
- [2.4 Lists](#section4)
- [2.5 Dictionaries](#section5)
- [2.6 Tuples](#section6)
- [2.7 Sets](#section7)
- [2.8 Other Data Types](#section8)
- [2.9 Appendix: String Formatting](#section9)
- [References](#section10)

For your reference, the figure below provides a brief overview of the data types in Python.

<img style="float: left; height:360px;" src="images/pic1.jpg">

# 2.1 Introduction<a id="section1"/>

Python is an interpreted, high-level, general-purpose programming language. Created by Guido van Rossum and first released in 1991, Python's design philosophy emphasizes code readability. Its language constructs and object-oriented approach aim to help programmers write clear, logical code for small- and large-scale projects. Python has become a common language for machine learning and data science applications. 

Python 3.0, released in 2008, was a significant revision of the language that is not entirely backward-compatible, and much Python 2 code does not run unmodified on Python 3.  This course makes use of Python 3. A non-profit organization, the Python Software Foundation (PSF), manages and directs resources for Python development. On January 1, 2020, the PSF discontinued the Python 2 language and no longer provides security patches and other improvements. Python interpreters are available for many operating systems. 

### Dynamic Typing

Python uses *dynamic typing*, meaning you can reassign variables to different data types. This makes Python very flexible in assigning data types; it differs from other languages that are *statically typed*.

In [1]:
my_dogs = 2

In [2]:
# Show
my_dogs

2

In [3]:
# Reassign to a list
my_dogs = ['Sammy', 'Frankie']

In [4]:
my_dogs

['Sammy', 'Frankie']

**Pros of Dynamic Typing**
* Very easy to work with
* Faster development time

**Cons of Dynamic Typing**
* May result in unexpected bugs
* You need to be aware of the `type()` of objects

### Assigning Variables

Listed are several rules for variable names in Python.
* Names can not start with a number
* Names can not contain spaces, use _ instead
* Names can not contain any of these symbols  :'",<>/?|\!@#%^&*~-+
* It's considered best practice ([PEP8](https://www.python.org/dev/peps/pep-0008/#function-and-variable-names)) that names are lowercase letters with underscores
* Avoid using Python built-in keywords like `list` and `str`
* Avoid using the single characters `l` (lowercase letter el), `O` (uppercase letter oh) and `I` (uppercase letter eye) as they can be confused with `1` and `0`

Variable assignment follows `name = object`, where a single equals sign `=` is an **assignment operator**.

In [5]:
# Assign the integer object 5 to the variable name a
a = 5

In [6]:
a

5

Variables can be re-assigned.

In [7]:
a = 10
a

10

Python also lets you reassign variables with a reference to the same object.

In [8]:
a = a + 10
a

20

Python lets you use shortcuts to add, subtract, multiply and divide numbers with reassignment using `+=`, `-=`, `*=`, and `/=`.

In [9]:
a += 10
a

30

In [10]:
a *= 2
a

60

### Determining variable type with `type()`

Objects in Python usually have **built-in functions**. These functions can perform actions on the object.

For instance, we can check what type of object is assigned to a variable using Python's built-in `type()` function. 

In [11]:
type(a)

int

In [12]:
a = (1,2)

In [13]:
type(a)

tuple

It is also important to note that the double equal `==` operator is used to test the equality of two expressions. The single equal `=` operator is only used to assign values to variables in Python. <br>Testing for inequality is performed with the not equal `!=` operator. <br>The greater than `>`, less than `<`, greater than or equal `>=`, less than or equal `<=` all perform as would generally be accepted. 

# 2.2 Numbers<a id="section2"/>

Python has several various types of number objects. 

**Integers** are whole numbers, positive or negative. For example: 2 and -2.

**Floating point** numbers in Python have a decimal point, or use an exponential (e) to define the number. For example 2.0 and -2.1 are examples of floating point numbers. 4E2 (4 times 10 to the power of 2) is also an example of a floating point number in Python.

Other types of number objects that are less frequently used inlcude **complex numbers** have real and imaginary parts (e.g., 3+4j), **decimal numbers** have control over the precision and rounding of numbers (e.g., `Decimal('0.1')`), **fractions** are rational numbers with numerator and denominator (e.g., `Fraction(1,3)`).

### Basic Arithmetic Operations

In [14]:
# Addition
2+1

3

In [15]:
# Subtraction
2-1

1

In [16]:
# Multiplication
2*2

4

In [17]:
# Division
3/2

1.5

In [18]:
# Floor Division
7//4

1

The `//` operator (two forward slashes) truncates the decimal without rounding, and returns an integer result.

If we just want the remainder after division, we use the `%` modulo operator.

In [19]:
# Modulo (remainder)
7%4

3

In [20]:
# Powers (exponentiation)
2**3

8

In [21]:
# Can also do roots this way
4**0.5

2.0

In [22]:
# Order of Operations followed in Python
2 + 10 * 10 + 3

105

In [23]:
# Can use parentheses to specify orders
(2+10) * (10+3)

156

Floating-point numbers are implemented in computer hardware as binary fractions (0 and 1). Many decimal fractions cannot be accurately represented as binaryu fractions. For example, the decimal number 0.1 results in an infinitely long binary fraction of 0.000110011001100110011... and our computer only stores a finite number of decimal places. This will only approximate 0.1 but never be equal. Hence, it is the limitation of our computer hardware and not an error in Python.

In [24]:
# Display issue in python, due to using binary data to represent all data
f3 = 0.1 + 0.1 + 0.1 
f3

0.30000000000000004

In [25]:
# One solution to that is to use decimal numbers, since they have rounding mechanisms to obtain exact representations
from decimal import Decimal
f4 = Decimal('0.1') + Decimal('0.1') + Decimal('0.1') 
f4

Decimal('0.3')

If number types are mixed, Phython will do the conversion. 

In [26]:
# Mix int and float; python will convert int to float first 
a = 1 + 2.5 
print(a)
type(a)

3.5


float

In [27]:
# Logic comparison
5<3

False

In [28]:
# Convert between different types
a = 2
b = float(a)
type(b)

float

In [29]:
c = int(b)
type(c)

int

### Build-in mathematical functions
Examples are: pow, abs, round, type
The functions are built into Python interpreter. We do not need to import any packages. Check the list of all built-in functions in Python:
https://docs.python.org/3.10/library/functions.html

In [30]:
# Power (exponentiation)
pow(2,4)

16

In [31]:
round(3.006)

3

In [32]:
# Absolute value
abs(-3.4) 

3.4

### Python Modules for Numerical Operations
We can import Python modules, such as `math`, `random` to perform mathematical operations. 
https://docs.python.org/3.10/library/math.html#module-math

In [33]:
import math

math.floor(3.006)

3

In [34]:
import random

# Return a random flaoting point number in the range 0-1
r = random.random()
print(r)

0.08829086647051909


In [35]:
# Check documentation for help about built-in functions
help(pow) 

Help on built-in function pow in module builtins:

pow(x, y, z=None, /)
    Equivalent to x**y (with two arguments) or x**y % z (with three arguments)
    
    Some types, such as ints, are able to use a more efficient algorithm when
    invoked using the three argument form.



# 2.3 Strings <a id="section3"/>

> A ***string*** is an immutable sequence containing letters, words, and other characters.

Strings are used in Python to record text information, such as names. Strings in Python are *sequences*, which means that Python keeps track of every element in the string, and we use indexing to get particular elements.

### Creating a String
To create a string in Python we can use either single quotes or double quotes.

In [36]:
# Single word
'hello'

'hello'

In [37]:
# Entire phrase 
'This is also a string'

'This is also a string'

In [38]:
# We can also use double quote
"String built with double quotes"

'String built with double quotes'

Note that the code below shows an error, because the single quote in <code>I'm</code> stopped the string. 

In [39]:
# Be careful with quotes!
' I'm using single quotes, but this will create an error'

SyntaxError: invalid syntax (<ipython-input-39-da9a34b3dc31>, line 2)

You can use combinations of double and single quotes to get the complete statement.

In [40]:
"Now I'm ready to use the single quotes inside a string!"

"Now I'm ready to use the single quotes inside a string!"

### Printing a String

Using Jupyter notebook with just a string in a cell will automatically output strings, but the correct way to display strings in your output is by using a **print** function.

In [41]:
# We can simply declare a string
'Hello World'

'Hello World'

In [42]:
# Note that we can't output multiple strings this way; only the last string is displayed
'Hello World 1'
'Hello World 2'

'Hello World 2'

We can use a print statement to print a string, or mulitple strings in a cell.

In [43]:
print('Hello World 1')
print('Hello World 2')
print('Use \n to print a new line') # \n prints a new line
print('\n')
print('See what I mean?')

Hello World 1
Hello World 2
Use 
 to print a new line


See what I mean?



We can also use the built-in function **len()** to check the length of a string. It counts all of the characters in the string, including spaces and punctuation.

In [44]:
len('Hello World')

11

### String Indexing
Since strings are sequences, Python can use indexes to call parts of the sequence.

**Indexing** starts at 0 for Python. 

In [45]:
# Assign s as a string
s = 'Hello World'

In [46]:
#Check
s

'Hello World'

In [47]:
# Print the object
print(s) 

Hello World


In [48]:
type(s)

str

In [49]:
# Show first element
s[0]

'H'

Use a <code>:</code> to perform **slicing** which grabs everything up to a designated point.

In [50]:
# Grab everything past the first term all the way to the length of s which is len(s)
s[1:]

'ello World'

In [51]:
# Note that there is no change to the original s
s

'Hello World'

In [52]:
# Grab everything UP TO the 3rd index
s[:3]

'Hel'

The above slicing doesn't include the 3rd index. In Python, statements are usually in the context of "up to, but not including".

In [53]:
#Everything
s[:]

'Hello World'

We can also use negative indexing to go backwards.

In [54]:
# Last letter (one index behind 0 so it loops back around)
s[-1]

'd'

In [55]:
# Grab everything but the last letter
s[:-1]

'Hello Worl'

We can also use index and slice notation to grab elements of a sequence by a specified step size (the default is 1). For instance we can use two colons in a row and then a number specifying the frequency to grab elements. For example:

In [56]:
# Grab everything, but go in steps size of 1
s[::1]

'Hello World'

In [57]:
# Grab everything, but go in step sizes of 2
s[::2]

'HloWrd'

In [58]:
# We can use this to print a string backwards
s[::-1]

'dlroW olleH'

### String Properties
Strings are *immutable*. Once a string is created, the elements within it **can not** be changed or replaced. 

In [59]:
s

'Hello World'

In [60]:
# Let's try to change the first letter to 'x'
s[0] = 'x'

TypeError: 'str' object does not support item assignment

Something we can do is concatenate strings!

In [61]:
s

'Hello World'

In [62]:
# Concatenate strings!
s + ' concatenate me!'

'Hello World concatenate me!'

In [63]:
# We can reassign s completely
s = s + ' concatenate me!'

In [64]:
print(s)

Hello World concatenate me!


We can use the multiplication symbol to create repetition.

In [65]:
letter = 'z'
letter*10

'zzzzzzzzzz'

### Basic Built-in String Methods

Objects in Python can also have built-in methods. Methods are called with a period followed by the method name. Parameters are extra arguments we can pass into the method.

`object.method(parameters)`

Here are some examples of built-in methods in strings:

In [66]:
s

'Hello World concatenate me!'

In [67]:
# Upper Case a string
s.upper()

'HELLO WORLD CONCATENATE ME!'

In [68]:
# Lower case
s.lower()

'hello world concatenate me!'

In [69]:
# Split a string by blank space (this is the default)
s.split()

['Hello', 'World', 'concatenate', 'me!']

In [70]:
# Split by a specific element (doesn't include the element that was split on)
s.split('W')

['Hello ', 'orld concatenate me!']

In [71]:
# Check all buil-in methods for the string s
 dir(s)

['__add__',
 '__class__',
 '__contains__',
 '__delattr__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__getitem__',
 '__getnewargs__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__iter__',
 '__le__',
 '__len__',
 '__lt__',
 '__mod__',
 '__mul__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__rmod__',
 '__rmul__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 'capitalize',
 'casefold',
 'center',
 'count',
 'encode',
 'endswith',
 'expandtabs',
 'find',
 'format',
 'format_map',
 'index',
 'isalnum',
 'isalpha',
 'isdecimal',
 'isdigit',
 'isidentifier',
 'islower',
 'isnumeric',
 'isprintable',
 'isspace',
 'istitle',
 'isupper',
 'join',
 'ljust',
 'lower',
 'lstrip',
 'maketrans',
 'partition',
 'replace',
 'rfind',
 'rindex',
 'rjust',
 'rpartition',
 'rsplit',
 'rstrip',
 'split',
 'splitlines',
 'startswith',
 'strip',
 'swapcase',
 'title',
 'translate',
 'upper',
 'zfill']

Please see the [appendix](#section9) for information on string formatting. 

# 2.4 Lists<a id="section4"/>

> A ***list*** is a mutable ordered sequence of elements, written as a series of items within square brackets.

Lists can be thought of the most general version of a *sequence* in Python. Unlike strings, they are mutable, meaning the elements inside a list can be changed.

Lists are constructed with brackets `[]` and commas separating every element in the list.

In [72]:
# Assign a list to an variable named my_list
my_list = [1,2,3]

Lists can  hold different object types. For example:

In [75]:
my_list = ['A string',23,100.232,'o']
my_list

['A string', 23, 100.232, 'o']

Just like strings, the `len()` built-in function will tell you how many items are in the sequence of the list.

In [74]:
len(my_list)

4

### Indexing and Slicing
Indexing and slicing work just like in strings.

In [76]:
my_list = ['one','two','three',4,5]

# Grab element at index 0
my_list[0]

'one'

In [77]:
# Grab everything UP TO index 3
my_list[:3]

['one', 'two', 'three']

We can also use + to concatenate lists, just like we did for strings.

In [78]:
my_list + ['new item']

['one', 'two', 'three', 4, 5, 'new item']

Note: This doesn't actually change the original list.

In [79]:
my_list

['one', 'two', 'three', 4, 5]

You would have to reassign the list to make the change permanent.

In [80]:
# Reassign
my_list = my_list + ['add new item permanently']

In [81]:
my_list

['one', 'two', 'three', 4, 5, 'add new item permanently']

We can also use the * for a duplication method similar to strings:

In [82]:
# Make the list double
my_list * 2

['one',
 'two',
 'three',
 4,
 5,
 'add new item permanently',
 'one',
 'two',
 'three',
 4,
 5,
 'add new item permanently']

In [83]:
# Again doubling is not permanent
my_list

['one', 'two', 'three', 4, 5, 'add new item permanently']

Lists indexing will return an error if there is no element at that index. For example:

In [84]:
my_list[100]

IndexError: list index out of range

### Basic List Methods

There are parallels between arrays in another language and lists in Python. Lists in Python tend to be more flexible than arrays in other languages for two reasons: they have no fixed size (meaning we don't have to specify the size), and they have no fixed type constraint (like we've seen above).

Explained next are some special methods for lists:

In [85]:
# Create a new list
list1 = [1,2,3]

Use the **append** method to permanently add an item to the end of a list:

In [86]:
# Append
list1.append('append me!')

In [87]:
# Show
list1

[1, 2, 3, 'append me!']

Use **pop** to "pop off" an item from the list. By default pop takes off the last index, but you can also specify which index to pop off. 

In [88]:
# Pop off the 0 indexed item
list1.pop(0)

1

In [89]:
# Show
list1

[2, 3, 'append me!']

In [90]:
# Assign the popped element, remember default popped index is -1
popped_item = list1.pop()

In [91]:
popped_item

'append me!'

In [92]:
# Show remaining list
list1

[2, 3]

We can insert and remove elements from a list.

In [93]:
c = ['a', 'b', 'c']

# Insert at index 0
c.insert(0, 'a0')
print(c)

['a0', 'a', 'b', 'c']


In [94]:
# Remove
c.remove('b')
print(c)

['a0', 'a', 'c']


In [95]:
# Remove at index 0
del c[0]
print(c)

['a', 'c']


We can use the **sort** method and the **reverse** methods to also effect your lists:

In [96]:
new_list = ['a','e','x','b','c']

In [97]:
# Show
new_list

['a', 'e', 'x', 'b', 'c']

In [98]:
# Use reverse to reverse order (this is permanent!)
new_list.reverse()

In [99]:
new_list

['c', 'b', 'x', 'e', 'a']

In [100]:
# Use sort to sort the list (in this case alphabetical order)
new_list.sort()

In [101]:
new_list

['a', 'b', 'c', 'e', 'x']

In [102]:
# For numbers, sorting is in ascending order
list_of_numbers = [2, 4, 3, 7, 1]
list_of_numbers.sort()
list_of_numbers

[1, 2, 3, 4, 7]

Two lists can be combined into a single list by the **zip** function.

In [103]:
a = [1,2,3,4,5]
b = [5,4,3,2,1]

print(zip(a,b))

To see the results of the **zip** function, we convert the returned zip object into a list. As you can see, the **zip** function returns a list of tuples.  Each tuple represents a pair of items that the function zipped together.  The order in the two lists was maintained.

In [104]:
print(list(zip(a,b)))

[(1, 5), (2, 4), (3, 3), (4, 2), (5, 1)]


The usual method for using the **zip** command is inside of a for-loop.  The following code shows how a for-loop can assign a variable to each collection that the program is iterating. 

In [105]:
a = [1,2,3,4,5]
b = [5,4,3,2,1]

for x,y in zip(a,b):
    print(f'{x} - {y}')

1 - 5
2 - 4
3 - 3
4 - 2
5 - 1


Usually, both collections will be of the same length when passed to the **zip** command.  It is not an error to have collections of different lengths.  As the following code illustrates, the **zip** command will only process elements up to the length of the smaller collection.

In [106]:
a = [1,2,3,4,5]
b = [5,4,3]

print(list(zip(a,b)))

[(1, 5), (2, 4), (3, 3)]


### Nesting Lists
Python data structures support **nesting**, i.e., we can have data structures within data structures. For example: a list inside a list.

In [107]:
# Let's make three lists
lst_1=[1,2,3]
lst_2=[4,5,6]
lst_3=[7,8,9]

# Make a list of lists to form a matrix
matrix = [lst_1,lst_2,lst_3]

In [108]:
# Show
matrix

[[1, 2, 3], [4, 5, 6], [7, 8, 9]]

We can again use indexing to grab elements, but now there are two levels for the index. The items in the matrix object, and the items inside that list.

In [109]:
# Grab first item in matrix object
matrix[0]

[1, 2, 3]

In [110]:
# Grab first item of the first item in the matrix object
matrix[0][0]

1

### List Comprehensions
Python has an advanced feature **list comprehensions**. They allow for quick construction of lists.

In [111]:
# Build a list comprehension by deconstructing a for loop within a []
first_col = [row[0] for row in matrix]

In [112]:
first_col

[1, 4, 7]

In [113]:
# Even numbers in the range 0-10 
even_numbers = [x for x in range(10) if x % 2 == 0] 
even_numbers

[0, 2, 4, 6, 8]

In [114]:
# Squared elements in the even numbers list
even_squares = [x * x for x in even_numbers]
even_squares

[0, 4, 16, 36, 64]

# 2.5 Dictionaries <a id="section5"/>

> A ***dictionary*** is an unordered and mutable Python container that stores mappings of unique keys to values.

We've been learning about *sequences* in Python so far, and now we're going to learn about *mappings* in Python. If you're familiar with other languages you can think of these dictionaries as hash tables. 

Mappings are a collection of objects that are stored by a *key*, unlike a sequence that stores objects by their relative position. This is an important distinction, since mappings won't retain order since they have objects defined by a key.

Therefore, a Python dictionary consists of a collection of **keys** and associated **values**. A colon `:` separates each key from its value. The keys must be unique, and can appear only once in a dictionary. Also, keys must be immutable objects, such as strings, integers, floats, tuples. The associated values can be almost any Python object, and there are no restrictions.

Dictionaries are mutable, therefore the elements can be changed, added, and removed. 

### Constructing a Dictionary

In [115]:
# Make a dictionary with {} and : to signify a key and a value
my_dict = {'key1':'value1','key2':'value2'}

In [116]:
my_dict

{'key1': 'value1', 'key2': 'value2'}

In [117]:
# Call values by their key
my_dict['key2']

'value2'

Dictionaries are very flexible in the data types they can hold. For example:

In [118]:
# The values can be any object type
my_dict = {1:101,2:102, 3:103}

# Call an item
my_dict[2]

102

In [119]:
my_dict = {'key1':123,'key2':[12,23,33],'key3':['item0','item1','item2']}

# Let's call items from the dictionary
my_dict['key3']

['item0', 'item1', 'item2']

In [120]:
# Can call an index on that value
my_dict['key3'][0]

'item0'

In [121]:
# Can even call methods on that value
my_dict['key3'][0].upper()

'ITEM0'

We can change the values of a key as well. For instance:

In [122]:
# Subtract 123 from the value
print(my_dict['key1'])
my_dict['key1'] = my_dict['key1'] - 100

123


In [123]:
#Check
my_dict['key1']

23

We can also create keys by assignment. For instance, if we start with an empty dictionary, we can continually add to it:

In [124]:
# Create a new dictionary
d = {}

In [125]:
# Create a new key through assignment
d['animal'] = 'Dog'

In [126]:
# Can do this with any object
d['answer'] = 42

In [127]:
#Show
d

{'animal': 'Dog', 'answer': 42}

### Nesting with Dictionaries

Python has flexibility of nesting objects and calling methods on them. Let's see a dictionary nested inside a dictionary.

In [128]:
# Dictionary nested inside a dictionary nested inside a dictionary
d = {'key1':{'nestkey':{'subnestkey':32}}}

In [129]:
# Keep calling the keys
d['key1']['nestkey']['subnestkey']

32

### Dictionary Built-In Methods

There are a several built-in methods we can call on a dictionary. 

In [132]:
# Create a typical dictionary
d = {'key1':1,'key2':2,'key3':3}

In [133]:
# Method to return a list of all keys 
d.keys()

dict_keys(['key1', 'key2', 'key3'])

In [134]:
# Method to return all values
d.values()

dict_values([1, 2, 3])

# 2.6 Tuples <a id="section6"/>

> A ***tuple*** is a collection of objects which is ordered and immutable, and it is commonly written as a series of items in parentheses.

In Python, tuples are very similar to lists, with the main difference being that tuples are *immutable* sequences, unlike lists that are *mutable* sequences. Tuples are created similar to lists, but with `()` instead of `[]`.

The basic characteristics of tuples include:
- They are ordered collections of objects - like lists and strings, tuples are positionally ordered collections of objects (i.e., they are sequences), that maintain a left-to-right order among their contents.
- Are accessed by offset - like strings and lists, items in a tuple are accessed by positional offset (not by key); therefore, they support indexing and slicing.
- Tuples are immutable sequences - like strings and lists, tuples are sequences. However, unlike lists that are *mutable* sequences, tuples are *immutable* sequences (meaning they can not be changed in-place). 
- Are fixed-length, heterogeneous, and arbitrarily nestable - because tuples are immutable, their size cannot be changed (without making a new copy). Tuples can hold any type of object, including other compound objects (e.g., lists, dictionaries, other tuples), and hence, they support arbitrary nesting.
- Tuples are arrays of object references - like lists, tuples are best thought of as arrays of references (pointers) to other objects with allocated memory.


### Constructing Tuples

Tuples are constructed by using parentheses `()` with the items separated by commas. For example:

In [135]:
# Creating a tuple
t = (1, 2, 3)
t

(1, 2, 3)

In [136]:
# Check the length of the tuple using len(), just like a list
len(t)

3

In [137]:
# We can also mix object types: e.g., strings, integer numbers, floating-point numbers
t = ('one', 2, 490.2)

# Show
t

('one', 2, 490.2)

In [138]:
# Tuples, lists, or dictionaries can be nested into other tuples
w = ('one', 'two', (4, 5), 6, ['r', 100])
w

('one', 'two', (4, 5), 6, ['r', 100])

In [139]:
# An empty tuple
u = () 
u

()

In [140]:
# A 1-item tuple
v = ('thing', )    
v

('thing',)

Note that for a single-item tuple we need to place a comma after the item, that is, we use `(item,)` and not `(item)`, since parentheses can also be used to enclose expressions like `(1 + 2) * 3` = `9`.

In [141]:
# Note that the output of this cell is not a tuple
# The displayed output of the cell is not in parentheses, and it is an integer number, not a tuple
a = (3)
a

3

The built-in function `type()` allows to check the type of an object.

In [142]:
# The type of the variable a is integer number
print(type(a))

<class 'int'>


In [143]:
# This is a tuple
b = (3,)
b

(3,)

In [144]:
# The type of the variable b is tuple
print(type(b))

<class 'tuple'>


In [145]:
# Not a tuple
(1 + 4) * 3

15

In [146]:
# This is a tuple: note that (1+4,) is the same as (5,), and when multiplied by 3, the tuple is repeated 3 times
(1 + 4,) * 3

(5, 5, 5)

The parentheses `()` can be omitted in the syntax, and tuples can be created just by listing items separated with commas. Although the parentheses are mostly optional with tuples, there are a few cases when using parentheses is required, e.g., within a function call, or when nested in a larger expression. For beginners, it is recommended to always use parentheses, in order to avoid the above exceptions, and because they improve the code readability.

In [147]:
t = 'one', 2, 490.2
t

('one', 2, 490.2)

In [148]:
# A tuple with one item can be created just by adding a comma after the item without using parentheses
'hello',

('hello',)

### Sequencing Operations

Since tuples are positionally ordered collections of objects like strings and lists, indexing and slicing work for tuples.

In [149]:
t

('one', 2, 490.2)

In [150]:
# Use indexing just like in lists and strings
t[0]

'one'

In [151]:
t[1]

2

In [152]:
t[-1]

490.2

In [153]:
# Slicing
t[0:2]

('one', 2)

Other sequence operations, such as concatenation and repetition, are also supported for tuples, in a similar way as for lists and strings.

In [154]:
# Concatenation
(1, 'book') + ('notes', 4) 

(1, 'book', 'notes', 4)

In [155]:
# Repetition
(1, 'thing') * 4 

(1, 'thing', 1, 'thing', 1, 'thing', 1, 'thing')

Because tuples are sequences, we can also use `for` loop iterations and list comprehensions to print the elements of tuples. 

In [156]:
# Consider the following tuple
x = ('b', 'u', 'i', 'l', 'd', 'i', 'n', 'g')
x

('b', 'u', 'i', 'l', 'd', 'i', 'n', 'g')

In [157]:
# We can use a `for` loop iteration to print each of the items of the tuple on a separate line
for i in x:
    print(i)

b
u
i
l
d
i
n
g


In [158]:
# A list comprehension can also be used to print each of the items of the tuple x on a separate line
l = [print(i) for i in x]

b
u
i
l
d
i
n
g


### Built-in Methods for Tuples

Tuples have built-in methods, but not as many as lists do. Tuples do not have methods such as append(), remove(), extend(), insert() and pop() due to its immutable nature.

In [159]:
# Show
t

('one', 2, 490.2)

In [160]:
# Use .index to enter an item and return the index
t.index('one')

0

In [161]:
# Use .count to count the number of times a value appears
t.count('one')

1

In [162]:
# Count the number of times 2 appears in the tuple
u = (1, 2, 3, 2, 1, 2)
u.count(2)

3

### Immutability

To emphasize one more time that tuples are immutable, check the following examples.

In [163]:
# If we try to change the first element, we will get an error message
t[0] = 'four'

TypeError: 'tuple' object does not support item assignment

Because of the immutability, tuples can't grow. Once a tuple is created, we can not add to it.

In [164]:
# We will get an error message
t.append('nope')

AttributeError: 'tuple' object has no attribute 'append'

We can, however, make a new tuple based on a current tuple.

In [165]:
t = (t[0], 7, t[2])
t

('one', 7, 490.2)

### Conversion to Lists

Conversions to lists and back to tuples is straightforward.

In [166]:
# Tuple to list
l = list(t)
l

['one', 7, 490.2]

In [167]:
# List to tuple
l2 = ['aa', 'bb', 5, 'cc']
t2 = tuple(l2)
t2

('aa', 'bb', 5, 'cc')

### Tuple Unpacking

Tuple unpacking means pairing objects on the right side of the assignment operator `=` with targets on the left side by position, and assigning them from left to right.

In [168]:
# Unpacking the tuple into the inidividual items
y = ('GOOG', 120, 490.2)
order, shares, price = y
print(order)
print(shares)
print(price)
print('Cost:', shares * price)

GOOG
120
490.2
Cost: 58824.0


In [169]:
# Unpacking the tuple: two names are entered for a tuple with 3 items, reuslting in an error
order, shares = y

ValueError: too many values to unpack (expected 2)

### Named Tuples

Named tuples are an extended type of tuples that allow items to be accessed by both position and attribute name, similar to dictionaries. They are created by using the `namedtuple` function from the collections module.

In [170]:
# Import and create a named tuple
from collections import namedtuple 
Rec = namedtuple('Rec', ['name', 'age', 'jobs']) 
# Assign a named-tuple record
bob = Rec(name='Bob', age=40.5, jobs=['dev', 'mgr']) 
bob

Rec(name='Bob', age=40.5, jobs=['dev', 'mgr'])

In [171]:
# Access by position
bob[0]

'Bob'

In [172]:
bob[1]

40.5

In [173]:
# Access by attribute 
bob.name, bob.jobs 

('Bob', ['dev', 'mgr'])

A named tuple can be converted to a dictionary, which allows key-based access to the items.

In [174]:
D = bob._asdict() 
# Access by key
D['name']

'Bob'

### When to Use Tuples

Although tuples are very similar to lists, tuples are not used as often as lists in programming. However, tuples are used when immutability is necessary; for instance, if in your program you are using an object and need to make sure it does not get changed, then a tuple provides convenient integrity. 

# 2.7 Sets <a id="section7"/>

>  A ***set*** is a collection of *unique* objects which is unordered and immutable, and are constructed by using the set() function.

Sets support operations corresponding to mathematical set theory (intersection, union, etc.). By definition, an item appears only once in a set, no matter how many times it is added. 

Because sets are collections of objects, they share some behavior with lists and dictionaries. For example, sets are iterable, can grow and shrink on demand, and may contain a variety of object types. 

However, since sets are unordered and do not map keys to values, they are neither a sequence or mapping type.

Sets have a variety of applications, especially in numeric and database-focused work.

To create a set object, pass in a sequence or another iterable object to the built-in set function.

In [175]:
x = set('abcde')
x

{'a', 'b', 'c', 'd', 'e'}

The sets are displayed with curly brackets. This is similar to a dictionary, but sets do not have keys and values (or, they can be considered dictionaries with only keys and without any values).

### Set Expressions

In [176]:
x = set('abcde')
y = set('bdxyz')
# Union
x | y

{'a', 'b', 'c', 'd', 'e', 'x', 'y', 'z'}

In [177]:
# Union
x | y

{'a', 'b', 'c', 'd', 'e', 'x', 'y', 'z'}

In [178]:
# Intersection
x & y

{'b', 'd'}

In [179]:
# Difference
x - y

{'a', 'c', 'e'}

In [180]:
# Symmetric difference (XOR) - elements in either x or y, but not both in x and y
x ^ y

{'a', 'c', 'e', 'x', 'y', 'z'}

In [181]:
# Superset, subset
x > y, x < y

(False, False)

In [182]:
# Membership of a set
'e' in x

True

Sets can also be created by adding elements to an existing set object.

In [183]:
# Create a set
z = set()

In [184]:
# Add to set with the add() method
z.add(1)
z

{1}

In [185]:
# Add a different element
z.add(2)
z

{1, 2}

In [186]:
# Try to add the same element
z.add(1)
z

{1, 2}

We cannot add another 1, because a set has only unique elements.

For instance, we can cast a list with multiple repeat elements into a set to get the unique elements of the list.

In [187]:
# Create a list with repeats
list1 = [1, 1, 2, 2, 3, 4, 5, 6, 1, 1]
# Cast as set to get unique values
set(list1)

{1, 2, 3, 4, 5, 6}

### Set methods

Similar to the set expressions shown above, there are set methods for union, intersections, and other related operations.

In [188]:
x = set('abcde')
y = set('bdxyz')
# Same as x & y
z = x.intersection(y) # Same as x & y
z

{'b', 'd'}

In [189]:
# Delete one item
z.remove('b') 
z

{'d'}

Also, we can use for-loops with the elements of sets.

In [190]:
for item in set('abc'):
    print(item * 3)

ccc
aaa
bbb


### Frozen Sets

Because sets are immutable, they don't support embedding mutable (a.k.a, not hashable) objects, like lists and dictionaries. 

In [191]:
# We get an error, because we cannot insert a list into a set
S = {3, 4}
S.add([1, 2, 3])

TypeError: unhashable type: 'list'

The `frozenset` built-in function creates an immutable set that cannot change and thus can be embedded in other sets. Frozen means that it freezes a mutable iterable object and makes it unchangeable.

In [192]:
s2 = frozenset([1,2,3])

In [193]:
s2

frozenset({1, 2, 3})

In [194]:
S.add(s2)
S

{3, 4, frozenset({1, 2, 3})}

# 2.8 Other Data Types <a id="section8"/>

### Booleans

Python also have a Boolean data type with predefined built-in name `True` and `False`, that are basically just the integers 1 and 0.

In [195]:
# Set object to be a boolean
a = True
# Show
a

True

The data type for True and False is `bool`.

In [196]:
print(type(a))

<class 'bool'>


We can also use comparison operators to create Booleans. 

In [197]:
# Output is boolean
1 > 2

False

In [198]:
# Is True the same as 1
True == 1

True

In Python each object is either True or False, as follows:
- Numbers are false if zero, and true otherwise.
- Other objects are false if empty, and true otherwise.

In [199]:
bool(2)

True

In [200]:
bool(0)

False

In [201]:
bool('book')

True

In [202]:
bool('')

False

In [203]:
bool([1, 2])

True

### The None Object

We can use **None** as a placeholder for an object that we don't want to reassign yet:

In [204]:
# None placeholder
b = None

In [205]:
# Show
print(b)

None


For instance, to initialize a list which size is not known yet, we can use None to preset the initial size and allow for future index assignment.


# 2.9 Appendix (not required for quizzes and assignments) <a id="section9"/>

# String Formatting

String formatting lets you inject items into a string, rather than trying to chain items together using commas or string concatenation. As a quick comparison, consider:

    player = 'Thomas'
    points = 33
    
    'Last night, '+player+' scored '+str(points)+' points.'  # concatenation
    
    f'Last night, {player} scored {points} points.'          # string formatting
    
    # The output of both concatenatin and string formatting is the same:
    'Last night Thomas scored 33 points.'


There are three ways to perform string formatting.
* The oldest method involves placeholders using the modulo `%` character.
* An improved technique uses the `.format()` string method.
* The newest method, introduced with Python 3.6, uses formatted string literals, called *f-strings*.

These three methods are described next each of them here.

### Formatting with placeholders
You can use <code>%s</code> to inject strings into your print statements. The modulo `%` is referred to as a "string formatting operator".

In [206]:
print("I'm going to inject %s here." %'something')

I'm going to inject something here.


You can pass multiple items by placing them inside a tuple after the `%` operator.

In [207]:
print("I'm going to inject %s text here, and %s text here." %('some','more'))

I'm going to inject some text here, and more text here.


You can also pass variable names:

In [208]:
x, y = 'some', 'more'
print("I'm going to inject %s text here, and %s text here."%(x,y))

I'm going to inject some text here, and more text here.


It should be noted that two methods <code>%s</code> and <code>%r</code> convert any python object to a string using two separate methods: `str()` and `repr()`. Where `%r` and `repr()` deliver the *string representation* of the object, including quotation marks and any escape characters.

In [209]:
print('He said his name was %s.' %'Fred')
print('He said his name was %r.' %'Fred') # note that 'Fred' is displayed in quotations

He said his name was Fred.
He said his name was 'Fred'.


As another example, `\t` inserts a tab into a string.

In [210]:
print('I once caught a fish %s.' %'this \tbig')
print('I once caught a fish %r.' %'this \tbig')

I once caught a fish this 	big.
I once caught a fish 'this \tbig'.


The `%s` operator converts whatever it sees into a string, including integers and floats. The `%d` operator converts numbers to integers first. Note the difference below:

In [211]:
print('I wrote %s programs today.' %3.75)
print('I wrote %d programs today.' %3.75)   

I wrote 3.75 programs today.
I wrote 3 programs today.


#### Padding and Precision of Floating Point Numbers
Floating point numbers use the format <code>%5.2f</code>. Here, <code>5</code> is the minimum number of characters the string should contain; these may be padded with whitespace if the entire number does not have this many digits. Next to this, <code>.2f</code> stands for how many numbers to show past the decimal point. 

In [212]:
print('Floating point numbers: %5.2f' %(13.144))

Floating point numbers: 13.14


In [213]:
print('Floating point numbers: %1.0f' %(13.144))

Floating point numbers: 13


In [214]:
print('Floating point numbers: %1.5f' %(13.144))

Floating point numbers: 13.14400


In [215]:
# Note that 5 empty spaces will be added in front of the number, to make it a total of 10 characters
print('Floating point numbers: %10.2f' %(13.144))

Floating point numbers:      13.14


In [216]:
print('Floating point numbers: %25.2f' %(13.144))

Floating point numbers:                     13.14


It is possible to use more than one conversion tool in the same print statement:

In [217]:
print('First: %s, Second: %5.2f, Third: %r' %('hi!',3.1415,'bye!'))

First: hi!, Second:  3.14, Third: 'bye!'


### Formatting with the `.format()` method
A better way to format objects into your strings for print statements is with the string `.format()` method. The syntax is:

    'String here {} then also {}'.format('something1','something2')
    
For example:

In [218]:
print('This is a string with an {}'.format('insert'))

This is a string with an insert


The `.format()` method has several advantages over the `%s` placeholder method:

1. Inserted objects can be called by index position:

In [219]:
print('The {2} {1} {0}'.format('fox','brown','quick'))

The quick brown fox


2. Inserted objects can be assigned keywords:

In [220]:
print('First Object: {a}, Second Object: {b}, Third Object: {c}'.format(a=1,b='Two',c=12.3))

First Object: 1, Second Object: Two, Third Object: 12.3


3. Inserted objects can be reused, avoiding duplication:

In [221]:
print('A %s saved is a %s earned.' %('penny','penny'))
# vs.
print('A {p} saved is a {p} earned.'.format(p='penny'))

A penny saved is a penny earned.
A penny saved is a penny earned.


Within the curly braces you can assign field lengths, left/right alignments, rounding parameters and more.

In [222]:
# The field 0 has a length of 8 characters, and the next field 1 has a length of 9 characters
print('{0:8} | {1:8}'.format('Fruit', 'Quantity'))
print('{0:8} | {1:8}'.format('Apples', 3.))
print('{0:8} | {1:8}'.format('Oranges', 10))

Fruit    | Quantity
Apples   |      3.0
Oranges  |       10


By default, `.format()` aligns text to the left, numbers to the right. You can pass an optional `<`,`^`, or `>` to set a left, center or right alignment:

In [223]:
print('{0:<8} | {1:^8} | {2:>8}'.format('Left','Center','Right'))
print('{0:<8} | {1:^8} | {2:>8}'.format(11,22,33))

Left     |  Center  |    Right
11       |    22    |       33


You can precede the aligment operator with a padding character.

In [224]:
print('{0:=<8} | {1:-^8} | {2:.>8}'.format('Left','Center','Right'))
print('{0:=<8} | {1:-^8} | {2:.>8}'.format(11,22,33))

Left==== | -Center- | ...Right
11====== | ---22--- | ......33


Field widths and float precision are handled in a way similar to placeholders. The following two print statements are equivalent:

In [225]:
print('This is my ten-character, two-decimal number:%10.2f' %13.579)
print('This is my ten-character, two-decimal number:{0:10.2f}'.format(13.579))

This is my ten-character, two-decimal number:     13.58
This is my ten-character, two-decimal number:     13.58


Note that there are 5 spaces following the colon, and 5 characters taken up by 13.58, for a total of ten characters.

### Formatted String Literals (f-strings)

Introduced in Python 3.6, `f-strings` offer several benefits over the older `.format()` string method described above. E.g., you can bring outside variables immediately into to the string rather than pass them as arguments through `.format(var)`.

In [226]:
name = 'Fred'

print(f"He said his name is {name}.")

He said his name is Fred.


Pass `!r` to get the string representation:

In [227]:
print(f"He said his name is {name!r}")

He said his name is 'Fred'


Float formatting follows `result: {value:{width}.{precision}}`

Where with the `.format()` method you might write `{value:10.4f}`, and with f-strings this can become `{value:{10}.{6}}`


In [228]:
num = 23.45678
print("My 10 character, four decimal number is:{0:10.4f}".format(num))
print(f"My 10 character, four decimal number is:{num:{10}.{6}}")

My 10 character, four decimal number is:   23.4568
My 10 character, four decimal number is:   23.4568


Note that with f-strings, *precision* refers to the total number of digits, not just those following the decimal. This fits more closely with scientific notation and statistical analysis. Unfortunately, f-strings do not pad to the right of the decimal, even if precision allows it:

In [229]:
num = 23.45
print("My 10 character, four decimal number is:{0:10.4f}".format(num))
print(f"My 10 character, four decimal number is:{num:{10}.{6}}")

My 10 character, four decimal number is:   23.4500
My 10 character, four decimal number is:     23.45


If this becomes important, you can always use `.format()` method syntax inside an `f-string`:

In [230]:
num = 23.45
print("My 10 character, four decimal number is:{0:10.4f}".format(num))
print(f"My 10 character, four decimal number is:{num:10.4f}")

My 10 character, four decimal number is:   23.4500
My 10 character, four decimal number is:   23.4500


# References <a id="section10"/>

1. Mark Lutz, "Learning Python," 5-th edition, O-Reilly, 2013. ISBN: 978-1-449-35573-9.
2. Pierian Data Inc., "Complete Python 3 Bootcamp," codes available at: [https://github.com/Pierian-Data/Complete-Python-3-Bootcamp](https://github.com/Pierian-Data/Complete-Python-3-Bootcamp).
3. Course T81 558:Applications of Deep Neural Networks, Washington University in St. Louis, Instructor: Jeff Heaton, codes available at: [https://github.com/jeffheaton/t81_558_deep_learning](https://github.com/jeffheaton/t81_558_deep_learning)